src/composables/protocol.ts

```diff
@@ -0,0 +1,56 @@
+import type {
+  Class,
+  IDefaultComposableOptions,
+  Protocol,
+} from '@/types';
+import { BaseProtocol } from '@/protocols/BaseProtocol';
+import { AeternityProtocol } from '@/protocols/aeternity/libs/AeternityProtocol';
+import { BitcoinProtocol } from '@/protocols/bitcoin/libs/BitcoinProtocol';
+import {
+  PROTOCOL_AETERNITY,
+  PROTOCOL_BITCOIN,
+} from '@/constants';
+import { watchUntilTruthy } from '@/utils';
+import { useAccounts } from './accounts';
+
+let isProtocolRegistryReady = true;
+
+const protocolClassReferences: Record<Protocol, Class<BaseProtocol> > = {
+  [PROTOCOL_AETERNITY]: AeternityProtocol,
+  [PROTOCOL_BITCOIN]: BitcoinProtocol,
+};
+const protocolInstanceRegistry: Partial<Record<Protocol, BaseProtocol>> = {};
+
+export function useProtocol({ store }: IDefaultComposableOptions) {
+  const { activeAccount } = useAccounts({ store });
+
+  function registerProtocol(protocol: Protocol, instance: BaseProtocol) {
+    if (protocolInstanceRegistry[protocol]) {
+      throw new Error('Protocol instance already exists in the registry');
+    }
```

Comment:
I think there's too much logic here. Why not just create the `getProtocolInstance` without using the `registerProtocol` method at all? Eg.:
```
function getProtocolInstance(targetProtocol?: Protocol): BaseProtocol {
  const protocol = targetProtocol || activeAccount.value.protocol;
  const registeredInstance = protocolInstanceRegistry[protocol];
  if (!registeredInstance) {
    const classReference = protocolClassReferences[protocol];
    if (classReference) {
      protocolInstanceRegistry[protocol] = new protocolClassReferences[protocol]();
      return protocolInstanceRegistry[protocol];
    } else {
      throw new Error('...');
    }
  }
  return registeredInstance;
}
```
Not tested, but in general skips the possibility of trying to register protocol that already is instantiated.

Reply:
https://github.com/aeternity/superhero-wallet/pull/2193/files#diff-b04f42206a6a2e9b4688bc42b5193f5cf351c98c98f28610bc7f842a6e07489b
Please have a look here